core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -2265,6 +2265,26 @@ final class Stream[+F[_], +O] private[fs2] (private val free: FreeC[F, O, Unit])
   def pauseWhen[F2[x] >: F[x]: Concurrent](pauseWhenTrue: Signal[F2, Boolean]): Stream[F2, O] =
     pauseWhen(pauseWhenTrue.discrete)
 
+  /** Returns a stream of zero or one elements. When the first element of this stream becomes
+    * available, the resulting stream emits a tuple consisting of that element and a new stream
+    * that emits the same elements as this stream, with the peeked element memoized.
+    *
+    * @example {{{
+    * scala> val outer =  Stream(1, 2, 3).peek.toList.head
+    * outer: (Int, Stream[Pure, Int]) = (1, Stream(..))
+    * scala> outer._2.toList
+    * res1: List[Int] = List(1, 2, 3)
+    * }}}
+    */
+  def peek: Stream[F, (O, Stream[F, O])] =
```

Comment:
Perhaps `uncons1` would be a better name for this. That name indicates that you are deconstructing the head and tail of the list (contrary to what a `cons` does), and is similar to what a similar method for `Pull` does.

Reply:
From `peek1` on ToPull:

`/** Like [[uncons1]] but does not consume the element (i.e., the element is pushed back). */`

It makes sense to call it `uncons1`, IMO. With the `1` because we don't emit the whole chunk.

Reply:
In this case, `peek1` seems like the right name. One thing I'm wary about is scoping bugs -- e.g., take a look at #1679 and look at similarity of the `uncons1` function.

Reply:
I was thinking the same, I think it would be nice to define clearer rules around this: i.e. do we permit to return the tail of the Stream from a `Pull` (and #1679 is a bug) or do we not?